sdl_android/src/main/java/com/smartdevicelink/api/audio/AudioStreamManager.java

```diff
@@ -0,0 +1,271 @@
+        if (streamingStateMachine.getState() != StreamingStateMachine.NONE) {
+            return;
+        }
+
+        try {
+            AudioPassThruCapabilities capabilities = (AudioPassThruCapabilities)internalInterface.getCapability(SystemCapabilityType.PCM_STREAMING);
+
+            switch (capabilities.getSamplingRate()) {
+                case _8KHZ:
+                    sdlSampleRate = 8000;
+                    break;
+                case _16KHZ:
+                    sdlSampleRate = 16000;
+                    break;
+                case _22KHZ:
+                    // common sample rate is 22050, not 22000
+                    // see https://en.wikipedia.org/wiki/Sampling_(signal_processing)#Audio_sampling
+                    sdlSampleRate = 22050;
+                    break;
+                case _44KHZ:
+                    // 2x 22050 is 44100
+                    // see https://en.wikipedia.org/wiki/Sampling_(signal_processing)#Audio_sampling
+                    sdlSampleRate = 44100;
+                    break;
+            }
+
+            switch (capabilities.getBitsPerSample()) {
+                case _8_BIT:
+                    sdlSampleType = SampleType.UNSIGNED_8_BIT;
+                    break;
+                case _16_BIT:
+                    sdlSampleType = SampleType.SIGNED_16_BIT;
+                    break;
+            }
+
+            streamingStateMachine.transitionToState(StreamingStateMachine.READY);
+
+            internalInterface.addServiceListener(SessionType.PCM, this.serviceListener);
+            internalInterface.startAudioService(encrypted);
+        } catch (Exception e) {
+            Log.e(TAG, "error starting audio stream", e);
+            streamingStateMachine.transitionToState(StreamingStateMachine.ERROR);
+        }
+    }
+
+    /**
+     * Stops the audio service and audio stream to the connected device.
+     * The method is non-blocking.
+     */
+    public void stopAudioStream() {
+        if (internalInterface == null || !internalInterface.isConnected()) {
+            return;
+        }
+
+        // streaming state must be STARTED (starting the service is ready. starting stream is started)
+        if (streamingStateMachine.getState() != StreamingStateMachine.STARTED) {
+            return;
+        }
+
+        streamingStateMachine.transitionToState(StreamingStateMachine.STOPPED);
+
+        internalInterface.stopAudioService();
+    }
+
+    /**
+     * Pushes the specified audio file to the playback queue.
+     * The audio file will be played immediately. If another audio file is currently playing
+     * the specified file will stay queued and automatically played when ready.
+     * @param audioFile The specified audio file to be played.
+     * @param completionListener A completion listener that informs when the audio file is played.
+     */
+    public void pushAudioFile(File audioFile, final CompletionListener completionListener) {

```

Comment:
@theRealBrett Not sure if a class like SdlFile is needed here. The purpose of that class is to upload a file using PutFile. Only Uri or Resource (as an AssetFileDescriptor) are useful but all other variables are ignored. For both useful variables I need a context. Does it make sense to confuse the developer with ignored stuff? Which one of the variables should I prefer if both are set? How should I get the context object?

I think it makes more sense to method overloads (one for for resource, another for Uri). This looks to me more clear regarding API design.

Reply:
I guess the one thing that I liked about `SdlFile`, or more what I was going for was letting the developer know the formats we supported and trying to keep some consistency since for pretty much every other file we use in the managers. Upon inspection of iOS, I see that they simply pass in the URL of the file to be played. Do you think that would be better here? Instead of having the developer implement the methods to obtain the byte arrays of file data? I guess that will need you to obtain context as well though. We can pass context in from the `SdlManager` to obtain it though like we did in the `FileManager`

Reply:
> Instead of having the developer implement the methods to obtain the byte arrays of file data?

I don't understand this...

Because of iOS I proposed the overload with Uri. Do you consider using Uri also for resources? I would still do the method overload for a resource Id and try https://stackoverflow.com/questions/7976141/get-uri-of-mp3-file-stored-in-res-raw-folder-in-android

Anyway I need a context. I can take a look at the file manager as you suggest. I kind of prefer to ask for a context using a parameter (due to context lifecycle)... any downsides from your point of view?

Reply:
I think Brett was suggesting that the AudioStreamManager could implement methods similar to these 

https://github.com/smartdevicelink/sdl_android/blob/feature/issue_782_Android_Managers/sdl_android/src/main/java/com/smartdevicelink/api/FileManager.java#L360

https://github.com/smartdevicelink/sdl_android/blob/feature/issue_782_Android_Managers/sdl_android/src/main/java/com/smartdevicelink/api/FileManager.java#L384

so that the developer can simply provide the resource id or Uri of the desired file and the manager converts it to a File object. I was using this method when testing this PR:

```
private File getSampleFile(int id) {
		File file = new File(getCacheDir() + "/" + "sample");

		try {
			InputStream is = getResources().openRawResource(id);
			FileOutputStream fos = new FileOutputStream(file);
			byte[] buffer = new byte[1 << 12];
			int length;
			do {
				length = is.read(buffer);
				if (length > 0) {
					fos.write(buffer, 0, length);
				}
			} while (length > 0);

			fos.flush();
			fos.close();
			is.close();
		} catch (IOException e) {
			e.printStackTrace();
		}

		return file;
	}
```

`FileManager` takes in a `Context` in its constructor, and holds a `WeakReference` to prevent a memory leak.